### PR TITLE
`Assessment`: Fix several issues with example submissions

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

In discussion with @fh314 several bugs related to example submissions were mentioned / shown. 

#### Modeling Exercises:
- Student explanation field missing in assessment view of example submission. Fixed in Commit: ad990b004253255bf6882d52b694518af6ea980e
- References to grading criteria are lost when using an existing submission as an example submission. Fixed in Commit: ebcd76d2ea4203de4404bbe772f9264e46df7fe8
#### Text Exercises:
- References to grading criteria are lost when using an existing submission as an example submission. Fixed in Commit: 0fa46d042f949187636362f009d7f1444f64b205 
- References to grading criteria in an example submission are lost when importing an text exercise. Fixed in Commit: a7e4f58a802b8b7bd150ac4bc85cc40e9d5637dd
- Lazy initialization error when importing text exercises with text blocks. Fixed in Commit: f4fb7fabfe4667d51246d450ea3daeec43f6f14b
- Existing assessment of an example submission is deleted without warning if user clicks on "edit submission". Fixed in Commit: 86eb0fa16eef2ea3b506da5320cb36519db9867a by adding an extra confirmation step (see screenshot)
- Import of text exercise **and** "use as example submission" are broken with internal server error if a **automatically** created text block and a **manually created** textblock by the assessor start with the same index in the assessed text (Duplicate key Error): Fixed in e76698659d322f4ad87e2484193d8acb4fd90b7d
- When importing an text exercise or using "use as a example submission" on a submission, the text block type is lost and all text blocks are now wrongly classified as manually created by an assessor: Fixed in: 7c8c501c938b87577e757d2f9bb36d7380260d99
- Automatic Blocks are not generated and displayed for example submissions. Fixed in 65f4c14e7b4e0b348710769eecd3d6d09536bde3
- Assessment analytics were sent for example-submissions causing an internal server error because of missing participation id. Fixed in: 586187bff04d9d4a068984a96c207eb3ee33a680

### Steps for Testing

**In my opinion it is not helpful to test that the above mentioned bugs do not show up any more, as I have confirmed this myself.**

Test Routine for Modeling and Text Example Submission:
- Create a example submission + assessment completely from scratch  as instructor and review as tutor
- Import an existing reviewed student submission as example submission + assessment as instructor and review as tutor

### Review Progress

# EACH FIX IS IN A SEPARATE COMMIT: FOR EASY REVIEW JUST GO COMMIT BY COMMIT

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
- Added or Modified Integration tests to check that the above mentioned bugs are fixed.

### Screenshots
**Explanation Box now showed**
image
** Warning before user edits the submission of an text exercise **
image

### CODE REVIEW

**Review by ge65cer — CHANGES_REQUESTED** [review]
When assessing an example submission, a request is sent to the server (`https://artemis-test1.artemis.in.tum.de/api/analytics/text-assessment/events`) that returns with an HTTP error status code of 400 (Bad Request).

Screenshot 2022-08-25 at 13 57 56

**Review by canberkanar — DISMISSED** [review]
Tested on TS1 with text and modeling exercises. All feedback and links to grading instructions within feedback are preserved correctly. Nice fix 👍🏻

**Review by akesfeden — DISMISSED** [review]
Tested on TS1 during testing session. Text and modeling exercises are successfully saved as example submissions with grading criteria preserved. Warning prompt when user clicks on "edit submission" works as expected. We noticed an issue about analytics events which will be handled in either this or a follow-up PR so I am approving now.

**Review by manuelmanso — APPROVED** [review]
Tested in TS1 as part of the testing session and works as expected, code also looks good.

**Comment by stefanwaldhauser** [comment]
> When assessing an example submission, a request is sent to the server (`https://artemis-test1.artemis.in.tum.de/api/analytics/text-assessment/events`) that returns with an HTTP error status code of 400 (Bad Request).
> 
> Screenshot 2022-08-25 at 13 57 56

Good catch! I fixed this in 586187bff04d9d4a068984a96c207eb3ee33a680

File: src/main/webapp/app/exercises/text/assess/analytics/text-assesment-analytics.service.ts
Comment (akesfeden) [inline src/main/webapp/app/exercises/text/assess/analytics/text-assesment-analytics.service.ts:75]:
```suggestion
        return !!this.location?.path().includes('example-submission');
```
Just a minor suggestion to make it more concise.

**Review by akesfeden — APPROVED** [review]
Re-tested in TS1 to confirm analytics events are not sent anymore, looks good. Added a minor code change suggestion.

**Review by akesfeden — APPROVED** [review]
Thanks for incorporating my suggestion. Re-approved.

**Review by manuelmanso — APPROVED** [review]
Re-approve after fixing the analytics error

File: src/test/javascript/spec/service/text-assessment-analytics.service.spec.ts
Comment (ge65cer) [inline src/test/javascript/spec/service/text-assessment-analytics.service.spec.ts:17]: Codacy warning.

**Review by ge65cer — APPROVED** [review]
Thanks for addressing my feedback, approved.
Please remove the unused imports before we merge this.

**Pull request merged**
